src/MongoDB.Driver/Core/Clusters/Cluster.cs

```
@@ -306,21 +240,91 @@ public ICoreSessionHandle StartSession(CoreSessionOptions options)
-            return string.Format(
-                "A timeout occurred after {0}ms selecting a server using {1}. Client view of cluster state is {2}.",
-                ms.ToString(),
-                selector.ToString(),
-                clusterDescription.ToString());
+            _serverSelectionEventLogger.LogAndPublish(new ClusterSelectingServerEvent(
+                clusterDescription,
+                selector,
+                EventContext.OperationId,
+                EventContext.OperationName));
+
+            var allSelectors = new List<IServerSelector>(5);
+            if (Settings.PreServerSelector != null)
+            {
+                allSelectors.Add(Settings.PreServerSelector);
+            }
+
+            allSelectors.Add(selector);
+            if (Settings.PostServerSelector != null)
+            {
+                allSelectors.Add(Settings.PostServerSelector);
+            }
+
+            allSelectors.Add(_latencyLimitingServerSelector);
+            var operationCountSelector = new OperationsCountServerSelector(Array.Empty<IClusterableServer>());
+            allSelectors.Add(operationCountSelector);
+
+            return (new CompositeServerSelector(allSelectors), operationCountSelector, Stopwatch.StartNew());
+        }
+
+        private void EndServerSelection(ClusterDescription clusterDescription, IServerSelector selector, ServerDescription selectedServerDescription, Stopwatch stopwatch)
+        {
+            stopwatch.Stop();
+            _serverSelectionEventLogger.LogAndPublish(new ClusterSelectedServerEvent(
+                clusterDescription,
+                selector,
+                selectedServerDescription,
+                stopwatch.Elapsed,
+                EventContext.OperationId,
+                EventContext.OperationName));
+        }
+
+        public Exception HandleServerSelectionException(ClusterDescription clusterDescription, IServerSelector selector, Exception exception, Stopwatch stopwatch)
```

looks to me like an internal detail so can this be private?
Also given the name of the method, is it better if this method takes care of throwing the exception instead of returning the exception? what do you think?

---

Changed to private, thanks! Cannot make the method throw, as it will confuse compiler and it will think we should return something from the method, as there will be no guaranty that `HandleServerSelectionException` throws.